### DISCUSSION TITLE
참조타입 값들은 어떻게 여러 타입이 가능한가요?

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#36] [데이터 타입]
- 참조타입 값들은 어떻게 여러 타입이 가능한가요?
  - C언어의 경우, 배열에는 특정 타입의 데이터만 들어가는데요, 이러한 방식의 장점은 무엇인가요?

**Comment by pipisebastian** [comment]
>참조타입 값들은 어떻게 여러 타입이 가능한가요

참조타입의 값, 예를 들어 배열에 들어가는 원소들은, 원시타입처럼 값을 직접 저장하는게 아니라, 값이 저장된 메모리 주소를 참조합니다. (참조타입의 배열이 메모리 주소를 참조하는 것 뿐만 아니라, 배열의 원소들도 메모리 주소를 참조. 한단계를 더 거치는 것)

즉, 배열 요소는 메모리 주소를 참조하므로, 서로 다른 타입의 데이터를 저장할 수 있습니다.

>C언어의 경우, 배열에는 특정 타입의 데이터만 들어가는데요, 이러한 방식의 장점은 무엇인가요?

메모리 관리가 효율적입니다. 
자바스크립트는, 참조타입에 무슨 값이 들어갈지 모르기 때문에, 여유로운 메모리 확보가 필요합니다. 반면 c언어는 모든 요소가 동일한 크기이기에, 미리 예측이 가능하여 효율적인 관리가 가능합니다.

**Comment by publdaze** [comment]
> - 참조타입 값들은 어떻게 여러 타입이 가능한가요?
- 메모리 공간 크기가 런타임에 결정되기 때문에 타입 체크 후 메모리 할당이 가능
> - C언어의 경우, 배열에는 특정 타입의 데이터만 들어가는데요, 이러한 방식의 장점은 무엇인가요?
- 예측이 쉽고 메모리 공간의 크기가 정해져 있음
